### PR TITLE
fix(settings): Support `order` property on App Discover elements and hide future elements

### PR DESCRIPTION
**Pull request by susnux**
## Summary

This adds support to pinning elements by setting the `order` property on the element (e.g. `order: 0` will always be the first element to show). Also filter list of elements to remove upcoming and outdated elements (as the json might be cached).

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/settings/src/components/AppStoreDiscover/AppStoreDiscoverSection.vue [hunk 1]
```diff
@@ -50,7 +50,7 @@ const elements = ref<IAppDiscoverElements[]>([])
  * Shuffle using the Fisher-Yates algorithm
  * @param array The array to shuffle (in place)
  */
-const shuffleArray = (array) => {
+const shuffleArray = <T, >(array: T[]): T[] => {
```
Comment (Pytal) [inline apps/settings/src/components/AppStoreDiscover/AppStoreDiscoverSection.vue:53]:
```suggestion
const shuffleArray = <T>(array: T[]): T[] => {
```

**Review by Pytal — APPROVED** [review]

File: apps/settings/src/components/AppStoreDiscover/AppStoreDiscoverSection.vue (hunk 1 above)
Comment (susnux) [reply apps/settings/src/components/AppStoreDiscover/AppStoreDiscoverSection.vue:53]: This does not work for lambda functions

**Review by artonge — APPROVED** [review]

File: apps/settings/src/utils/appDiscoverParser.ts [hunk 2]
```diff
@@ -45,3 +45,25 @@ export const apiTypeParser = (element: Record<string, unknown>): IAppDiscoverEle
...
+	if (element.date) {
+		// Element not yet published
+		if (element.date - now > 0) {
+			return false
+		}
+	}
+
+	if (element.expiryDate) {
+		// Element expired
+		if (element.expiryDate - now < 0) {
```
[the remark this answers is not in the document]
Comment (susnux) [reply apps/settings/src/utils/appDiscoverParser.ts:64]: yes!

**Pull request merged**